designs/0003-parallel_autodiff.md

```diff
@@ -0,0 +1,705 @@
+the same scalar type and the summation must be associative as the
+summation order is not necessarily guaranteed.
+
+Note that the actual slicing of the input data is not performed by the
+reduce function itself. Instead, the actual sub slicing is left to the
+user provided function which knows best how to do that (and slicing is
+implemented in Stan itself already, so we should not duplicate this in
+math). In practice, it makes most sense for Stan programs to use
+counting iterators. Doing so will make the binary function receive
+starting and end indexes (given as integers) which represent the sub
+slice of data to do the actual reduction over. The actual data
+involved then must be part of the closure of the function object
+BinaryFunction passed into the facility.
+
+In c++ code using the `parallel_reduce_sum` may look as:
+
+```c++
+// reduce functor which is the BinaryFunction
+// here we use iterators which represent integer indices
+template <typename T>
+struct count_lpdf {
+  const std::vector<int>& data_;
+  const T& lambda_;
+
+  count_lpdf(const std::vector<int>& data, const T& lambda)
+      : data_(data), lambda_(lambda) {}
+
+  // does the reduction in the sub-slice start to end
+  inline T operator()(std::size_t start, std::size_t end) const {
+    const std::size_t elems = end - start + 1;
+    std::vector<int> partial_data;
+    partial_data.insert(partial_data.end(), data_.begin() + start,
+                        data_.begin() + end + 1);
+    return stan::math::poisson_lpmf(partial_data, lambda_);
+  }
+};
+
+TEST(parallel_reduce_sum, gradient) {
+  double lambda_d = 10.0;
+  const std::size_t elems = 10000;
+  const std::size_t num_iter = 1000;
+  std::vector<int> data(elems);
+  
+  for (std::size_t i = 0; i != elems; ++i)
+    data[i] = i;
+
+  typedef boost::counting_iterator<std::size_t> count_iter;
+  using stan::math::var;
+
+  var lambda = lambda_d;
+
+  count_lpdf<var> reduce_op(data, lambda);
+
+  var poisson_lpdf = stan::math::parallel_reduce_sum(
+      count_iter(0), count_iter(elems), var(0.0), reduce_op);
+
+  var lambda_ref = lambda_d;
+  var poisson_lpdf_ref = stan::math::poisson_lpmf(data, lambda_ref);
+
+  EXPECT_FLOAT_EQ(value_of(poisson_lpdf), value_of(poisson_lpdf_ref));
+
+  stan::math::grad(poisson_lpdf.vi_);
+  stan::math::grad(poisson_lpdf_ref.vi_);
+
+  EXPECT_FLOAT_EQ(poisson_lpdf.adj(), poisson_lpdf_ref.adj());
+}
+```
+
+The above example already runs on a prototype of this RFC. For this
+example of $10^4$ Poisson log-likelihood evaluations we go from about
+320ms on 1 core down to 170ms on 2 cores and 150ms on 3 cores. With 4
+cores execution time increases again. This was executed on a macOS
+system and shows greatly the potential of the approach, I think.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The overall strategy to implement all of this is based on
+
+- Introduce the Intel TBB which provides a work stealing queuing
```

Comment:
It seems like Intel TBB is an orthogonal concern, isn't it? If so, it's important, but it's not critical to the design discussion.

Also, is the design incapable of working with plain threads?